### ISSUE TITLE
Problem getting classifire classifications (very slow execution and large number of failures)

### ISSUE THREAD

**Issue by mapio** [iomega/ms2query#222]
I was finally able to run the README example to build a new library. To avoid the problem of having not enough spectra I've used `GNPS-NIH-NATURALPRODUCTSLIBRARY.mgf` from https://external.gnps2.org/gnpslibrary that contains around 1200 unique SMILES.

The library construction took more than 6h; observing the execution, the slowest part amounted to the retrieval of classifire classifications. The execution reported around 800 failures with the progress meter reporting a speed of one minute per item (see the attached extract from the [attached log](https://github.com/iomega/ms2query/files/13448719/not_found.txt)).

A code inspection reveals that the classification are retrieved (in `ms2query/create_new_library/add_classifire_classifications.py`) by querying the URL `http://classyfire.wishartlab.com/entities/{full_inchikey}.json`. 

Such site seems to be very intermittently available, several tests on our side end up with a 504 HTTP error code and a very long response time, as this example shows:

```shell
$ time curl http://classyfire.wishartlab.com/entities/BBLGCDSLCDDALX-LKGBESRRSA-N.json
<html>
<head><title>504 Gateway Time-out</title></head>
<body>
<center><h1>504 Gateway Time-out</h1></center>
<hr><center>nginx</center>
</body>
</html>

real    1m0.255s
user    0m0.010s
sys     0m0.012s
```

I'd like to ask if it is common, in your experience, to have such an high rate of failures leading to a very high execution time. 

Moreover, since in our case another workflow step seeks the classifier information, it would be very useful to know if there is a safe way to skip such step in the construction of the library, or if such information is required to construct more accurate models.

**Comment by niekdejonge** [comment]
I noticed the same issue. This is due to the server of classyfire being down. Their website cannot be reached at all right now (and yesterday as well). This is not something on the MS2Query side. Unfortunately I have not found a way yet to use classyfire independent of their API. Probably/hopefully it will work fine again in a few days again. It is possible to create a model without adding the classyfire annotations, but this will mean you won't have the compound class annotations in your predictions.

**Comment by mapio** [comment]
Of course the failure is due to an external site, I was just curious to know if that was a common condition, in your experience.

But much more importantly, I'm happy to know that the classyfire annotations are not required (we will recover such information in another step of our workflow).

But I can't see how to prevent such step, can you please suggest the best way to proceed with library creation without querying classyfire? Is there a parameter, or configuration I can use, or should I edit the ms2qyery code (i.e. commenting the step)?

**Comment by niekdejonge** [comment]
It is not very common to be down, but I don't consider it very stable either. So in the past I looked into doing the classyfire calculations locally in the code instead, but as far as I could find they only offer the API option, which is a bit frustrating. If this changed, or if you found a way to run it locally, I would be interested :) 

You can set the add_compound_classes parameter in the LibraryFilesCreator to False. I checked and this is unfortunately not (yet) very easy to do directly in the clean_and_train_models function. I will look into quickly adding this option. For now you can add this to train_models.py line 80. So adding add_compound_classes =False.

**Comment by mapio** [comment]
I've prepared two PR to allow to propagate the parameter from the outer function to LibraryFilesCreator. I'm submitting them, please choose the one the fits better with your programming style.

**Comment by justinjjvanderhooft** [comment]
Thanks for letting us know that things are working out for you now, and for proposing solutions on how to deal with the library creation without classification terms. Indeed, some external services are not always responsive, and this is something outside our control. I do hope we can in the longer run replace it by something that we do have under control....

**Comment by mapio** [comment]
One thing that I found of great help, when depending on external APIs, is to cache the replies so that in case the workflow needs to be run again, at least the already obtained answers are found in the cache and can be very efficiently retrieved without imposing other load on the external service.

If you are interested in such an approach, I'll be glad to give you further details.

**Comment by niekdejonge** [comment]
@mapio Thanks a lot for creating the PRs!  I also already implemented the same  in #223, I will check and merge them in a minute. 

Good suggestion, for a different tool (matchms) I implement exactly the same approach for looking up compounds on pubchem. We could certainly do this here for the classifier lookup as well in the future.
